From ed94fefc05e69378ecc69f53b0c88979ea53ac60 Mon Sep 17 00:00:00 2001
From: Sami Shalayel <sami.shalayel@qt.io>
Date: Tue, 12 Aug 2025 08:15:20 +0200
Subject: [PATCH 106/239] qmlls.ini generation: check parent build dir before
 writing it

When we see a qt_add_qml_module(URI org.kde.foo) call, we assume during the
.qmlls.ini generation that the QML module will end up at <build>/org/kde/foo/.
When this assumption doesn't hold, for example when CMakeLists.txt is in
/some/path/to/source/CMakeLists.txt, we added /some/ as build directory
because of a missing check that "path", "to" and "source" are equal to "org",
"kde" and "foo".

Add a check and use the build directory as is when a mismatch between
uri and CML.txt is detected.

Fixes: QTBUG-138599
Change-Id: Ifba0086e70ca9a8f5c9212dd97fb5c8636875090
Reviewed-by: Ulf Hermann <ulf.hermann@qt.io>
(cherry picked from commit 8ba50d5f31772e1e696a1db51585a67b848525d8)
Reviewed-by: Qt Cherry-pick Bot <cherrypick_bot@qt-project.org>
---
 src/qml/Qt6QmlMacros.cmake                      | 17 +++++++++++++++--
 .../test_generate_qmlls_ini/CMakeLists.txt      |  2 +-
 .../WrongOutput/CMakeLists.txt                  |  7 +++++++
 .../WrongOutput/Main.qml                        |  3 +++
 .../auto/cmake/test_generate_qmlls_ini/main.cpp |  7 +++++++
 5 files changed, 33 insertions(+), 3 deletions(-)
 create mode 100644 tests/auto/cmake/test_generate_qmlls_ini/WrongOutput/CMakeLists.txt
 create mode 100644 tests/auto/cmake/test_generate_qmlls_ini/WrongOutput/Main.qml

diff --git a/src/qml/Qt6QmlMacros.cmake b/src/qml/Qt6QmlMacros.cmake
index 7cd6f4079e..0927ad82d7 100644
--- a/src/qml/Qt6QmlMacros.cmake
+++ b/src/qml/Qt6QmlMacros.cmake
@@ -1061,10 +1061,23 @@ Check https://doc.qt.io/qt-6/qt-cmake-policy-qtp0001.html for policy details."
             set(output_folder "${CMAKE_CURRENT_BINARY_DIR}")
         endif()
         string(REPLACE "." ";" uri_bits "${arg_URI}")
-        set(build_folder "${output_folder}")
+        list(REVERSE uri_bits)
+        set(build_folder_current "${output_folder}")
+        set(use_build_folder true)
         foreach(bit IN LISTS uri_bits)
-            get_filename_component(build_folder "${build_folder}" DIRECTORY)
+            get_filename_component(current_folder "${build_folder_current}" NAME)
+            get_filename_component(build_folder_current "${build_folder_current}" DIRECTORY)
+            if(NOT current_folder STREQUAL bit)
+                set(use_build_folder false)
+                break()
+            endif()
         endforeach()
+        if(use_build_folder)
+            set(build_folder "${build_folder_current}")
+        else()
+            set(build_folder "${output_folder}")
+        endif()
+
 
         if(QT_QML_GENERATE_QMLLS_INI)
             set_property(DIRECTORY APPEND PROPERTY _qmlls_ini_build_folders "${build_folder}")
diff --git a/tests/auto/cmake/test_generate_qmlls_ini/CMakeLists.txt b/tests/auto/cmake/test_generate_qmlls_ini/CMakeLists.txt
index eb43ceb682..aa5c0b6b86 100644
--- a/tests/auto/cmake/test_generate_qmlls_ini/CMakeLists.txt
+++ b/tests/auto/cmake/test_generate_qmlls_ini/CMakeLists.txt
@@ -18,7 +18,7 @@ add_subdirectory(Dependency)
 add_subdirectory(ModuleWithDependency)
 add_subdirectory(quotesInPath)
 add_subdirectory(QmllsBuildIni)
-
+add_subdirectory(WrongOutput)
 add_subdirectory(WithoutCMakeBuilds)
 
 qt_add_qml_module(tst_generate_qmlls_ini
diff --git a/tests/auto/cmake/test_generate_qmlls_ini/WrongOutput/CMakeLists.txt b/tests/auto/cmake/test_generate_qmlls_ini/WrongOutput/CMakeLists.txt
new file mode 100644
index 0000000000..dd7e1e7f8c
--- /dev/null
+++ b/tests/auto/cmake/test_generate_qmlls_ini/WrongOutput/CMakeLists.txt
@@ -0,0 +1,7 @@
+# Copyright (C) 2024 The Qt Company Ltd.
+# SPDX-License-Identifier: BSD-3-Clause
+
+qt_add_qml_module(wrongoutput
+    URI org.kde.foo
+    QML_FILES Main.qml
+)
diff --git a/tests/auto/cmake/test_generate_qmlls_ini/WrongOutput/Main.qml b/tests/auto/cmake/test_generate_qmlls_ini/WrongOutput/Main.qml
new file mode 100644
index 0000000000..3052615aef
--- /dev/null
+++ b/tests/auto/cmake/test_generate_qmlls_ini/WrongOutput/Main.qml
@@ -0,0 +1,3 @@
+import QtQuick
+
+Item {}
diff --git a/tests/auto/cmake/test_generate_qmlls_ini/main.cpp b/tests/auto/cmake/test_generate_qmlls_ini/main.cpp
index 073f3cf5e1..a06cbe7a91 100644
--- a/tests/auto/cmake/test_generate_qmlls_ini/main.cpp
+++ b/tests/auto/cmake/test_generate_qmlls_ini/main.cpp
@@ -114,6 +114,13 @@ void tst_generate_qmlls_ini::qmllsIniAreCorrect_data()
                 << withoutCMakeBuilds.absolutePath() << QStringList{ build.absolutePath(), }
                 << u"true"_s << docPath << defaultImportPaths;
     }
+    {
+        QDir wrongOutput = source;
+        QVERIFY(wrongOutput.cd(u"WrongOutput"_s));
+        QTest::addRow("wrong-output")
+                << wrongOutput.absolutePath() << QStringList{ build.filePath("WrongOutput"), }
+                << noCMakeCalls << docPath << defaultImportPaths;
+    }
 }
 
 void tst_generate_qmlls_ini::qmllsIniAreCorrect()
-- 
2.51.2

